pkg/protocol/engine/blockdag/inmemoryblockdag/blockdag.go

```diff
@@ -101,6 +121,20 @@ func (b *BlockDAG) Attach(data *model.Block) (block *blocks.Block, wasAttached b
 	if block, wasAttached, err = b.attach(data); wasAttached {
 		b.events.BlockAttached.Trigger(block)
 
+		// We add blocks that commit to a commitment we haven't committed ourselves yet to this limited size buffer and
+		// only let them become solid once we committed said slot ourselves (to the same commitment).
+		// This is necessary in order to make sure that all necessary state is available after a block is solid (specifically
+		// the state of the referenced commitment for the commitment filter). All the while, we need to make sure that
+		// the requesting of missing blocks (done in b.attach) continues so that we can advance our state and eventually
+		// commit the slot.
+		// This limited size buffer has a nice side effect: In normal behavior (e.g. no attack of a neighbor that sends you
+		// unsolidifiable blocks in your committed slots) it will prevent the node from storing too many blocks in memory.
```

Comment:
This limits the ability for a node to run out of memory without warpsync. Which is obviously a good thing as we will start to through away blocks that are committing to not yet committed slots. Since the buffer is an LRU cache, we will ultimately request the more recent blocks and should be able to eventually solidify.

Reply:
We have already concluded that this should be a compatible mechanism. In theory, when warpsync is there, there is not reason to keep the protocol-level buffer of blocks: as you will solidify the commitment chain, drop the corresponding block and delegate the solidification to warpsync instead of caching observed unsolid blocks directly.
We although decided to also keep the protocol buffer in place for now, as, without it, it will complicate syntetic tests that don't rely on a constant activity: in fact, if you throw away the block that triggered the initial commitment chain solidification, you will need to observe another one building on the (now solid) commit chain to trigger a warpsync against the matched engine's chain.